feat(databases-collections, compass-sidebar): add rename collection flow COMPASS-5704

baileympearson opened the pull request:
## Description

This PR adds a new rename collection flow to Compass.  The rename collection flow is behind a feature flag.

https://jira.mongodb.org/browse/COMPASS-7480 will ensure that all data associated with the old namespace is refreshed when the user renames a namespace.

### Checklist
- [x] New tests and/or benchmarks are included
- [x] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it's updating a dependancy, link to the Pull Request that originally introduced the fix -->
- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Open Questions
<!--- Any particular areas you'd like reviewers to pay attention to? -->

## Dependents
<!--- If applicable, link PRs/commits that this PR is dependent on or is a dependency of. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

Anemy reviewed (commented):
Still in draft, sorry for review, leaving a few comments on small things. I'll take a look tomorrow into Compass' handling of collections names and see if I have any suggestions for that TODO.
I'll also give this a run. 
Looking good! Big first Compass ticket 😅  Nice work figuring all the things out.

gribnoysup commented:
> unsure on scope

We don't have strict rules around this, but usually this is just the most relevant package / plugin name, or multiple ones. In your case in can be `feat(databases-collections, compass-sidebar)` or something similar

gribnoysup commented:
> properly refresh state throughout the app with new collection state when a collection is renamed

Hah, that's a challenging one! Some places are getting their information about collection from collection model, so in theory it should be as easy as updating collection model instance and making sure that all relevant places are subscribed to the change, but some places, in particular collection tabs are just pulling this informantion once from the model during instantiation and will not be aware if something changes on the model, so this ticket might require changing all those places that require collection information to use the model instead of just passing those values as props to the store so that we have consistency in where the data is being grabbed from.

> reselect the selected collection (if selected) after renaming

This I think is actually related to the previous one, I think re-selecting might not be the best approach here as this will require a lot of manual wiring of events to the palces that rely on this data, plus it might make it easy to skip less obvious cases, like inactive tabs for a namespace that is being renamed.

In addition to the TODOs you outlined, I wonder what are we planning to do about saved aggregations and queries for the renamed namespace, those should probably be updated too or do we want to keep them with the old namespace? 🤔

mcasimir commented on packages/compass-app-stores/src/stores/instance-store.js:
```diff
@@ -251,6 +251,17 @@ store.onActivated = (appRegistry) => {
+  appRegistry.on('collection-renamed', async ({ from, to }) => {
+    // we must fetch the old collection's metadata before refreshing because refreshing the
+    // collection metadata will remove the old collection from the model.
+    const metadata = await store.fetchCollectionMetadata(from);
```
q: what happens if this one fails? do we have some magic catch all handling on the `appRegistry` or the rejection would fly around?

gribnoysup commented on packages/compass-app-stores/src/stores/instance-store.js:
```diff
@@ -251,6 +251,17 @@ store.onActivated = (appRegistry) => {
+  appRegistry.on('collection-renamed', async ({ from, to }) => {
+    // we must fetch the old collection's metadata before refreshing because refreshing the
+    // collection metadata will remove the old collection from the model.
+    const metadata = await store.fetchCollectionMetadata(from);
```
Why are we fetching metadata for an old namespace and not a new one? Is it to avoid refreshing collection before we updated collection tabs?

baileympearson replied:
as far as I know, there's no way to wait for `refresh-data` to have refreshed the collections.  If we refresh the collection metadata in the instance store before selecting the collection metadata, it would be non-deterministic as to which collection metadata we get back - the stale collection metadata or the new collection metadata

so the flow is currently:
- fetch the old collection metadata
- let the instance store refresh
- in `tabs.ts`, construct new collection metadata by spreading the replacing the namespace in the old collection metadata

gribnoysup replied:
So yeah, while you can't await on those events, the main handling for it happens in this store, so if that's the only reason, you can just call and await `refreshInstance` method directly, similar to what the handler for this event is doing (with `refreshNamespace` probably being a better choice here so that we don't need to refresh stats for all namespaces, only for affected one)

I do think though that potentially what you're doing is the way to go just because as we discussed, compass doesn't really expects namespace to ever be changed, so refreshing the namespace will effectively leave a tab open for a non-existent namespace, so maybe better to update state first and then trigger a namespace refresh. That's what I was trying to confirm, I assumed you maybe tried it the other way around already.

gribnoysup commented on packages/compass-app-stores/src/stores/instance-store.js:
```diff
@@ -251,6 +251,17 @@ store.onActivated = (appRegistry) => {
+      metadata,
+      newNamespace: to,
+    });
+    appRegistry.emit('refresh-data');
```
This seems a bit drastic, we are always renaming in the same database namespace, right? I think `refreshNamespace` should be enough here

gribnoysup commented on packages/compass-collection/src/modules/tabs.ts:
```diff
@@ -182,56 +183,80 @@ const reducer: Reducer<CollectionTabsState> = (
 
+function createNewTabInternal(
+  collectionMetadata: CollectionMetadata,
+  dispatch: any,
```
If you need access to `dispatch`, I'd suggest turning this into a thunk action instead of passing it around without a proper type. Alternatively you can set up the interface of this method in a way that doesn't require you to pass it here, like passing a less specific onSubtabChange callback or something like (or not factoring this subscription in a shared method at all)

baileympearson replied:
`dispatch` is typed locally, I just haven't pushed those changes yet

gribnoysup commented on packages/compass-collection/src/modules/tabs.ts:
```diff
@@ -182,56 +183,80 @@ const reducer: Reducer<CollectionTabsState> = (
-    const localAppRegistry = new AppRegistry();
-    const store = collectionTabRole.configureStore({
-      dataService,
+    return createNewTabInternal(
```
Not actually sure completely why we moved it to a separate method. This `createNewTab` is already a shared action that returns a tab when dispatched, it's a bit unconventional, but the cleaenst way for now that gives us access to thunk provided services. Seems to me like both open collection and the new action that you added can just continue using it unless I'm missing something

baileympearson replied:
Here's my perhaps (incorrect) understanding.  `createNewTab` is a thunk action that opens a new tab with the given collection metadata at the end of the current set of tabs.  When a user renames a collection, we want to "refresh" the tabs that they have open for the old collection to have fresh collection data.  The easiest way to refresh the tabs is to close and reopen them.

If we simply dispatch a `createNewTab` for every tab that needs to be refreshed, we'd also have to emit corresponding `closeTab` actions for the old tab, _and_ we'd need to then reorder the tabs so that each newly opened tab is in the correct tab location.

I refactored this logic into a separate function because in my `collectionRenamed` thunk, I map over all the currently opened tabs and replace any that match the old namespace with a new tab.  That refreshes each stale collection tab while preserving the ordering without requiring multiple create/close actions + reordering of tabs.

gribnoysup commented on packages/compass-collection/src/modules/tabs.ts:
```diff
@@ -182,56 +183,80 @@ const reducer: Reducer<CollectionTabsState> = (
+  if (action.type === CollectionTabsActions.CollectionRenamed) {
+    const { tabs, activeTabIndex } = action;
+
+    const activeTabId = tabs[activeTabIndex]?.id;
```
Reducer should be responsible for deriving new state from the old one and action, so in this case instead of passing active tab index from the action, you can do:

```suggestion
const activeTabIndex = getActiveTabIndex(state);
const activeTabId = tabs[activeTabIndex]?.id ?? null;
```

gribnoysup reviewed (commented):
Awesome work! This is like the messiest part of Compass state, and you totally managed to deal with it!

There are two things I'd want to highlight in addition to the small comments I left here and there:

So as we don't really have a single source of current namespace state, most of the plugins rely on listening to particular events to update their internally kept reference of the active tab. We will be addressing this pretty soon, but for now we'd probably need to stick with the pattern we have for handling this. To keep those updated, other plugins listen to following events:

```
'select-database'
'select-namespace'
'open-instance-workspace'
'open-namespace-in-new-tab'
```

The tricky part here is that the currently active tab state is only known by the tabs store, so to close the loop, you might need to do something similar to what we do for dropped collections: when tabs plugin reacts to tabs refresh event, it might need to emit something like active-namespace-renamed afterwards so that other plugins like home and sidebar can react. Without this you loose the active tab selector in the sidebar and the app title is not updated (those are just the most noticeable things that caught my eye). We can take a look together if you feel like it!

Another thing is a question that we should raise with design / product (unless there is an answer for this already that you know): should something happen to saved and recent queries and saved aggregations when rename happens? Currently when you rename you don't see them anymore in those "Open saved" popovers as they are filtered by namespace

baileympearson commented on packages/databases-collections/src/components/rename-collection-modal/rename-collection-modal.tsx:
```diff
@@ -0,0 +1,145 @@
+      )}
+      {modalState === 'confirmation-screen' && (
+        <div data-testid="rename-collection-confirmation-screen">
+          {`Are you sure you want to rename "${initialCollectionName}" to "${newName}"?`}{' '}
```
I'm not sure, I think something might have gotten messed up in a reformt?  Fixed!

baileympearson commented on packages/databases-collections/src/components/rename-collection-modal/rename-collection-modal.tsx:
```diff
@@ -0,0 +1,145 @@
+          />
+        </FormFieldContainer>
+      )}
+      {modalState === 'confirmation-screen' && (
```
Yes, I ran the designs by Misba Friday and it's included in the changes I pushed

baileympearson commented:
> There are two things I'd want to highlight in addition to the small comments I left here and there:
> 
> So as we don't really have a single source of current namespace state, most of the plugins rely on listening to particular events to update their internally kept reference of the active tab. We will be addressing this pretty soon, but for now we'd probably need to stick with the pattern we have for handling this. To keep those updated, other plugins listen to following events:
> 
> 'select-database'
> 'select-namespace'
> 'open-instance-workspace'
> 'open-namespace-in-new-tab'
> The tricky part here is that the currently active tab state is only known by the tabs store, so to close the loop, you might need to do something similar to what we do for dropped collections: when tabs plugin reacts to tabs refresh event, it might need to emit something like active-namespace-renamed afterwards so that other plugins like home and sidebar can react. Without this you loose the active tab selector in the sidebar and the app title is not updated (those are just the most noticeable things that caught my eye). We can take a look together if you feel like it!
> 
> Another thing is a question that we should raise with design / product (unless there is an answer for this already that you know): should something happen to saved and recent queries and saved aggregations when rename happens? Currently when you rename you don't see them anymore in those "Open saved" popovers as they are filtered by namespace

@gribnoysup I introduced a feature flag into this PR to help break up the review.  I think we can focus on what's included here first, and then focus ensuring we update state with the new namespace.  If that sounds good to you, I'll create another ticket to address the two issues you've outlined here.

You'll notice that I did fix the saved aggregations bug. I fixed that Friday before I decided to add a feature flag

gribnoysup commented on packages/databases-collections/src/index.ts:
```diff
@@ -57,6 +61,19 @@ export const DropNamespacePlugin = registerHadronPlugin(
+  {
+    dataService:
+      dataServiceLocator as typeof dataServiceLocator<'renameCollection'>,
+    toastService: toastProviderLocator,
```
As I mentioned in another comment, let's not do that, services are only needed for platform specific stuff, you can import and call `openToast` directly, it's a method that works in every environment

gribnoysup commented on packages/compass-aggregations/src/modules/saved-pipeline.ts:
```diff
@@ -86,6 +86,39 @@ export const updatePipelineList =
       });
   };
 
+export const handleNamespaceChanged = ({
```
For the follow-up PRs: would it make more sense to do this in rename collection plugin instead of here? You'd also need to do it for saved queries and recent queries and spreading this logic across multiple plugins doesn't feel right
